File: library/evolution/src/main/java/cz/matfyz/evolution/category/SMO.java
package cz.matfyz.evolution.category;

import cz.matfyz.core.metadata.MetadataCategory;
import cz.matfyz.core.schema.SchemaCategory;
import cz.matfyz.evolution.exception.EvolutionException;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonSubTypes({
    @JsonSubTypes.Type(value = CreateObjex.class, name = "createObjex"),
    @JsonSubTypes.Type(value = DeleteObjex.class, name = "deleteObjex"),
    @JsonSubTypes.Type(value = UpdateObjex.class, name = "updateObjex"),
    @JsonSubTypes.Type(value = CreateMorphism.class, name = "createMorphism"),
    @JsonSubTypes.Type(value = DeleteMorphism.class, name = "deleteMorphism"),
    @JsonSubTypes.Type(value = UpdateMorphism.class, name = "updateMorphism"),
    @JsonSubTypes.Type(value = Composite.class, name = "composite"),
})
public interface SMO {

    <T> T accept(SchemaEvolutionVisitor<T> visitor) throws EvolutionException;

    void up(SchemaCategory schema, MetadataCategory metadata) throws EvolutionException;

    void down(SchemaCategory schema, MetadataCategory metadata) throws EvolutionException;

}

File: client/src/types/evocat/schema/smo.ts
import type { Category } from '@/types/schema';

export enum SMOType {
    CreateObjex = 'createObjex',
    DeleteObjex = 'deleteObjex',
    UpdateObjex = 'updateObjex',
    CreateMorphism = 'createMorphism',
    DeleteMorphism = 'deleteMorphism',
    UpdateMorphism = 'updateMorphism',
    Composite = 'composite',
}

export type SMOResponse<T extends SMOType = SMOType> = {
    type: T;
};

export type SMO<T extends SMOType = SMOType> = {
    readonly type: T;
    toServer(): SMOResponse<T>;
    up(category: Category): void;
    down(category: Category): void;
}
